crates/iceberg/src/puffin/metadata.rs

```diff
@@ -0,0 +1,809 @@
+        }
+    }
+
+    pub(crate) fn bit_number(&self) -> BitNumber {
+        match self {
+            Flag::FooterPayloadCompressed => BitNumber(0),
+        }
+    }
+
+    fn from(byte_and_bit: &(ByteNumber, BitNumber)) -> Option<Flag> {
+        FLAGS_BY_BYTE_AND_BIT.get(byte_and_bit).cloned()
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+/// Metadata about a puffin file.
+/// For more information, see: https://iceberg.apache.org/puffin-spec/#filemetadata
+pub(crate) struct FileMetadata {
+    /// Metadata about blobs in file
+    pub(crate) blobs: Vec<BlobMetadata>,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    #[serde(default)]
+    /// Arbitrary meta-information, like writer identification/version.
+    pub(crate) properties: HashMap<String, String>,
+}
+
+impl FileMetadata {
+    pub(crate) const MAGIC_LENGTH: u8 = 4;
+    pub(crate) const MAGIC: [u8; FileMetadata::MAGIC_LENGTH as usize] = [0x50, 0x46, 0x41, 0x31];
+
+    // We use the term FOOTER_STRUCT to refer to the fixed-length portion of the Footer, as illustrated below.
+    //
+    //                        Footer
+    //                          |
+    //  -------------------------------------------------
+    // |                                                 |
+    // Magic FooterPayload FooterPayloadLength Flags Magic
+    //                     |                             |
+    //                      -----------------------------
+    //                                    |
+    //                              FOOTER_STRUCT
+
+    const FOOTER_STRUCT_PAYLOAD_LENGTH_OFFSET: u8 = 0;
+    const FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH: u8 = 4;
+    const FOOTER_STRUCT_FLAGS_OFFSET: u8 = FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_OFFSET
+        + FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH;
+    pub(crate) const FOOTER_STRUCT_FLAGS_LENGTH: u8 = 4;
+    const FOOTER_STRUCT_MAGIC_OFFSET: u8 =
+        FileMetadata::FOOTER_STRUCT_FLAGS_OFFSET + FileMetadata::FOOTER_STRUCT_FLAGS_LENGTH;
+    pub(crate) const FOOTER_STRUCT_LENGTH: u8 =
+        FileMetadata::FOOTER_STRUCT_MAGIC_OFFSET + FileMetadata::MAGIC_LENGTH;
+
+    fn check_magic(bytes: &[u8]) -> Result<()> {
+        if bytes != FileMetadata::MAGIC {
+            Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Bad magic value: {:?} should be {:?}",
+                    bytes,
+                    FileMetadata::MAGIC
+                ),
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn read_footer_payload_length(
+        file_read: &dyn FileRead,
+        input_file_length: u64,
+    ) -> Result<u32> {
+        let start = input_file_length - u64::from(FileMetadata::FOOTER_STRUCT_LENGTH);
+        let end = start + u64::from(FileMetadata::FOOTER_STRUCT_PAYLOAD_LENGTH_LENGTH);
+        let footer_payload_length_bytes = file_read.read(start..end).await?;
```

Comment:
Please note that we usually need to avoid reading small bytes from cloud storage. Therefore, we will need something like prefetching, which will read a large amount of data capable of holding the entire footer. We can implement this optimization in a separate PR.

Reply:
Agreed that there is definitely room for optimization here which we can do in a separate PR and focus on correct-ness instead here. I've added a sub-task to the [ticket](https://github.com/apache/iceberg-rust/issues/744)  for now to keep track of this.